<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="UM Bioinformatics Core" />


<title>Advanced Snakemake</title>

<script src="site_libs/header-attrs-2.29/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/paper.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link rel="shortcut icon" href="favicon-16x16.png" />
<!--
Favicon from
https://twemoji.twitter.com/
https://favicon.io/emoji-favicons/repeat-button
-->
<link rel="apple-touch-icon" sizes="180x180" href="/apple-touch-icon.png">
<link rel="icon" type="image/png" sizes="32x32" href="/favicon-32x32.png">
<link rel="icon" type="image/png" sizes="16x16" href="/favicon-16x16.png">
<link rel="shortcut icon" href="favicon-16x16.png" />
<link rel="manifest" href="/site.webmanifest">
<link rel="mask-icon" href="/safari-pinned-tab.svg" color="#5bbad5">
<meta name="msapplication-TileColor" content="#da532c">
<meta name="theme-color" content="#ffffff">

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-inverse  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Reproducible Computing</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="intro.html">Intro</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Day 1
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Overview &amp; Quick-Start</a>
      <ul class="dropdown-menu" role="menu">
        <li>
          <a href="Module_overview_and_warmup.html">Overview and Warmup</a>
        </li>
        <li>
          <a href="Module_sneak_peek_great_lakes.html">Sneak Peek of Great Lakes</a>
        </li>
        <li>
          <a href="Module_data_priorities_analysis_setup.html">Data Priorities and Analysis Setup</a>
        </li>
      </ul>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">UMich Resources</a>
      <ul class="dropdown-menu" role="menu">
        <li>
          <a href="Module_storage_best_practices_UMRCP.html">Raw Data, Compute/Storage Options, &amp; UMRCP</a>
        </li>
        <li>
          <a href="Module_great_lakes_cluster.html">Great Lakes HPC</a>
        </li>
      </ul>
    </li>
    <li class="dropdown-submenu">
      <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">Software Management</a>
      <ul class="dropdown-menu" role="menu">
        <li>
          <a href="Module_compute_environment_definition.html">Terminology - Compute Environment</a>
        </li>
        <li>
          <a href="Module_software_management_conda.html">Software Management and Conda</a>
        </li>
      </ul>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Day 2
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module_containers_docker_singularity.html">Containerization with Docker/Singularity</a>
    </li>
    <li>
      <a href="Module_intro_to_workflow_automation.html">Intro to workflow automation</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Day 3
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Module_intro_to_snakemake.html">Intro to Snakemake</a>
    </li>
    <li>
      <a href="Module_advanced_snakemake.html">Advanced Snakemake</a>
    </li>
    <li>
      <a href="Module_transferring_data_globus.html">Transferring data with Globus</a>
    </li>
  </ul>
</li>
<li>
  <a href="wrap_up.html">Wrap up</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Advanced Snakemake</h1>
<h4 class="author">UM Bioinformatics Core</h4>

</div>


<style type="text/css">
body{ /* Normal  */
      font-size: 14pt;
  }
pre {
  font-size: 12pt
}

.wrapper { overflow-x: auto; }
.wrapper table { white-space: nowrap; }

</style>
<p>By the end of this module, we will:</p>
<ul>
<li>Show other ways to visualize the Snakemake workflow.</li>
<li>Use an <strong>all</strong> rule and <strong>wildcards</strong> to
easily extend a workflow to new inputs.</li>
<li>Introduce <strong>configuration files</strong></li>
<li>Run Snakemake on the HPC worker nodes instead of the login
nodes</li>
</ul>
<div id="visualizing-the-dag-rulegraph" class="section level2">
<h2>Visualizing the DAG / rulegraph</h2>
<p>In addition to doing a dry run, <code>snakemake</code> can generate a
diagram of the rules to execute. Snakemake thinks of the rules as a <a
href="https://en.wikipedia.org/wiki/Directed_acyclic_graph"
target="_blank">Directed Acyclic Graph</a> (DAG). You can see this with
the <code>--dag</code> flag and the <code>dot</code> program (part of <a
href="https://graphviz.org/" target="_blank">Graphviz</a>, an open
source graph visualizer).</p>
<pre class="sh"><code>snakemake --dag | dot -Tpng &gt; dag.png</code></pre>
<p><img src="images/advanced_snakemake/dag.png" alt="dag.png" />
<br/></p>
<p>As you can see above, by default <code>dot</code> displays the graphs
vertically with inputs at the top and outputs at the bottom. Our
Snakefile is arranged from output to input, so you can invert the
diagram like so:</p>
<pre class="sh"><code>snakemake --dag | dot -Grankdir=&quot;BT&quot; -Tpng &gt; dag_inverted.png</code></pre>
<p><img src="images/advanced_snakemake/dag_inverted.png"
alt="dag_inverted.png" /> <br/></p>
</div>
<div id="all-target-wildcards-visualizaing-the-rulegraph"
class="section level2">
<h2>All target / Wildcards / Visualizaing the rulegraph</h2>
<p>To emphasize the endpoint of the workflow and also clarify the
default target, it’s common to place a “rule all” at the top of the
file. (This also allows you to reorder rules within the file
arbitrarily.) It is unusual in that it only has an input directive.</p>
<pre><code>rule all:
    input: &quot;results/little_women_part_1.select_words.txt&quot;</code></pre>
<p>Snakemake <em>wildcards</em> make it easy to to extend the workflow
to new inputs. (And they typically reduce repetition of file names.)
Here is an excerpt of the Snakefile showing the original
<strong>select_words</strong> rule and the same rule with wildcards.</p>
<pre><code>rule all:
    input: &quot;results/little_women_part_1.select_words.txt&quot;

# original rule
# rule select_words:
#    input: &quot;result/little_women_part_1.sort_counts.txt&quot;
#    output: &quot;results/little_women_part_1.select_words.txt&quot;
#    shell: &quot;egrep &#39;^(Jo|Amy|Meg|Beth)\s&#39; {input} &gt; {output}&quot;

# now with wildcards
rule select_words:
    input: &quot;result/{base_name}.sort_counts.txt&quot;
    output: &quot;results/{base_name}.select_words.txt&quot;
    shell: &quot;egrep &#39;^(Jo|Amy|Meg|Beth)\s&#39; {input} &gt; {output}&quot;
</code></pre>
<p>How does this work?</p>
<ul>
<li>The <strong>all</strong> rule needs an input named
“results/little_women_part_1.select_words.txt”</li>
<li>The <strong>select_words</strong> rule can produce
“results/{base_name}.select_words.txt” which would match the
<strong>all</strong> input if the wildcard
<em>{base_name}</em>=“little_women_part_1”</li>
<li>In <strong>select_word</strong> input, the rule substitutes the
<em>{base_name}</em> value for it’s wildcard, making
“result/little_women_part_1.sort_counts.txt”</li>
<li>So the the value of the wildcard is determined by the file name in
the all rule and that value propagates from output to input.</li>
</ul>
<p>What happens if we run snakemake?</p>
<pre class="sh"><code>snakemake -c1 --dry-run</code></pre>
<blockquote>
<pre><code>Building DAG of jobs...
Nothing to be done (all requested files are present and up to date).</code></pre>
</blockquote>
<p>And that actually makes sense: even though we used wildcards, the
rules ended up with identical input and output values.</p>
<p>We can extend wildcards across remaining rules as below. Note that we
need to keep the literal value in the <strong>all</strong> rule so the
other rules can derive the value of the wildcard.</p>
<pre><code>rule all:
    input: &quot;results/little_women_part_1.select_words.txt&quot;

rule select_words:
    input: &quot;results/{base_name}.sort_counts.txt&quot;
    output: &quot;results/{base_name}.select_words.txt&quot;
    shell: &quot;egrep &#39;^(Jo|Amy|Meg|Beth)\s&#39; {input} &gt; {output}&quot;

rule sort_counts:
    input: &quot;results/{base_name}.count_words.txt&quot;
    output: &quot;results/{base_name}.sort_counts.txt&quot;
    shell: &quot;&quot;&quot;
    sort -k1,1nr {input} \
        | awk &#39;BEGIN {{print &quot;word\tcount&quot;}} {{ print $2 &quot;\t&quot; $1}}&#39; \
            &gt; {output}
        &quot;&quot;&quot;
rule count_words:
    input: &quot;results/{base_name}.split_words.txt&quot;
    output: &quot;results/{base_name}.count_words.txt&quot;
    shell: &quot;sort {input} | uniq -c &gt; {output}&quot;

rule split_words:
    input: &quot;inputs/{base_name}.txt&quot;
    output: &quot;results/{base_name}.split_words.txt&quot;
    shell: &quot;cat {input} | tr -cs &#39;[:alpha:]&#39; &#39;\n&#39; &gt; {output}&quot;</code></pre>
<p>Now that the Snakefile is using wildcards, it is trivial to extend it
to new inputs. We adjust the <strong>all</strong> rule adding a second
value, separated by a comma:</p>
<pre><code>rule all:
    input: &quot;results/little_women_part_1.select_words.txt&quot;, &quot;results/little_women_part_2.select_words.txt&quot;
...</code></pre>
<p>The addition of a new wildcard value (little_women_part_2) will
trigger all five rules.</p>
<pre class="sh"><code>snakemake --dry-run</code></pre>
<blockquote>
<pre><code>...
Job stats:
job             count
------------  -------
all                 1
count_words         1
select_words        1
sort_counts         1
split_words         1
total               5
...</code></pre>
</blockquote>
<p>Before we run this for real, let’s add a new flag
<code>--forceall</code> to tell snakemake to trigger every single rule
for all possible inputs. Basically act like there are no existing
outputs and it needs to start from scratch. But this is still a dry-run,
so no files will be changed.</p>
<pre class="sh"><code>snakemake --dry-run --forceall</code></pre>
<blockquote>
<pre><code>...
Job stats:
job             count
------------  -------
all                 1
count_words         2
select_words        2
sort_counts         2
split_words         2
total               9
...</code></pre>
</blockquote>
<p>Try generating a DAG visualization. (Now it’s truly looking more like
a graph.) Note:</p>
<ul>
<li>Each wildcard value gets it’s own path</li>
<li>The wildcard values are displayed at the input</li>
<li>Some box outlines are solid and other are dotted; why is that?</li>
</ul>
<pre class="sh"><code>snakemake --dag | dot -Grankdir=&quot;BT&quot; -Tpng &gt; dag_inverted_wildcards.png</code></pre>
<p><img src="images/advanced_snakemake/dag_inverted_wildcards.png"
alt="dag_inverted_wildcards.png" /> <br/></p>
<p>With complex workflows, you sometimes don’t want to see all the
possible inputs and would rather just see the abstract relationship of
the rules. Snakemake supports this with a related flag
<code>--rulegraph</code></p>
<pre class="sh"><code>snakemake --rulegraph | dot -Grankdir=&quot;BT&quot; -Tpng &gt; rulegraph.png</code></pre>
<p><img src="images/advanced_snakemake/rulegraph.png"
alt="rulegraph.png" /> <br/></p>
<p>Ok. Go ahead and run <code>snakemake</code> until you are 100%
complete and see these files in results:</p>
<blockquote>
<pre><code>$ ls results
little_women_part_1.count_words.txt   little_women_part_2.count_words.txt
little_women_part_1.select_words.txt  little_women_part_2.select_words.txt
little_women_part_1.sort_counts.txt   little_women_part_2.sort_counts.txt
little_women_part_1.split_words.txt   little_women_part_2.split_words.txt</code></pre>
</blockquote>
</div>
<div id="the-expand-function-config-files-parameters"
class="section level2">
<h2>The expand function / config files / parameters</h2>
<p>Ideally, a workflow could be run on a new set of inputs without
changing the Snakefile at all. For example, we should be able to use our
unmodified Snakefile with the input of of Shakespeare’s <a
href="https://www.gutenberg.org/ebooks/1513">Romeo and Juliet</a> to see
whether Montague is mentioned more than Capulet. Most of the rules in
the workflow are already general enough; but there are two
exceptions:</p>
<ul>
<li>The <strong>all</strong> rule mentions specific texts.</li>
<li>The <strong>select_words</strong> rule mentions specific names.</li>
</ul>
<p>We can extract these specifics from the Snakefile using a
<em>configuration file</em>. We’ll start with the
<strong>select_words</strong> rule.</p>
<pre><code>rule select_words:
    input: &quot;results/{base_name}.sort_counts.txt&quot;
    output: &quot;results/{base_name}.select_words.txt&quot;
    shell: &quot;egrep &#39;^(Jo|Amy|Meg|Beth)\s&#39; {input} &gt; {output}&quot;</code></pre>
<p>Firstly, we’ll pull the regular expression out of the shell directive
into a <a
href="https://snakemake.readthedocs.io/en/stable/snakefiles/configuration.html#configuration"
target="_blank">configuration file</a>. Snakemake supports several
styles of configuration file; we’ll use the YAML syntax. By convention
we’ll create config/config.yaml; you could create and edit this file
manually; alternatively you can also execute the line below:</p>
<pre class="sh"><code>mkdir config
echo &#39;select_words_regex: ^(Jo|Amy|Meg|Beth)\s&#39; &gt; config/config.yaml
cat config/config.yaml</code></pre>
<p>Now add this line above the <strong>all</strong> rule:</p>
<pre><code>configfile: &quot;config/config.yaml&quot;</code></pre>
<p>Add a parameter initialized to the config key
<em>select_words_regex</em>. Then adjust the select_words rule to
refernce that parameter.</p>
<pre><code>rule select_words:
    input: &quot;results/{base_name}.sort_counts.txt&quot;
    output: &quot;results/{base_name}.select_words.txt&quot;
    params: regex=config[select_words_regex]
    shell: &quot;egrep &#39;{params.regex}&#39; {input} &gt; {output}&quot;</code></pre>
<p>The <a
href="https://snakemake.readthedocs.io/en/stable/snakefiles/rules.html#non-file-parameters-for-rules"
target="_blank">params directive</a> is a way of adjusting the shell
command with a value that isn’t tied to the filesystem. Also note that
params can be named (e.g. <code>x="value"</code>) and referenced by name
in the shell (<code>{params.x}</code>).</p>
<p>Now that we’ve made this change, let’s dry-run snakemake. Note we’re
adding the <code>-p</code> flag so snakemake will print out the actual
commands it will execute:</p>
<pre class="sh"><code>snakemake --dry-run -p</code></pre>
<p>In the output, note:</p>
<ul>
<li>In Job stats, we see snakemake believes we need to rerun
<strong>select_words</strong> for both inputs.</li>
<li>In the rule details, the reason the rule is triggered is that code
and params have changed</li>
<li>Because we used <code>-p</code>, you can see the actual egrep
statement (which looks fine).</li>
</ul>
<blockquote>
<pre><code>Building DAG of jobs...
Job stats:
job             count
------------  -------
all                 1
select_words        2
total               3
.
[Mon Jun 10 23:01:21 2024]
rule select_words:
input: results/little_women_part_1.sort_counts.txt
output: results/little_women_part_1.select_words.txt
jobid: 1
reason: Code has changed since last execution; Params have changed since last execution
wildcards: base_name=little_women_part_1
resources: tmpdir=/tmp
.
egrep &#39;^(Jo|Amy|Meg|Beth)\s&#39; results/little_women_part_1.sort_counts.txt &gt; results/little_women_part_1.select_words.txt
.
[Mon Jun 10 23:01:21 2024]
rule select_words:
input: results/little_women_part_2.sort_counts.txt
output: results/little_women_part_2.select_words.txt
jobid: 5
reason: Code has changed since last execution; Params have changed since last execution
wildcards: base_name=little_women_part_2
resources: tmpdir=/tmp
.
egrep &#39;^(Jo|Amy|Meg|Beth)\s&#39; results/little_women_part_2.sort_counts.txt &gt; results/little_women_part_2.select_words.txt
...</code></pre>
</blockquote>
<p>Go ahead and run snakemake until 100% done.</p>
<p>Now we’ll pull the file base names from the <strong>all</strong> rule
into the config file:</p>
<pre><code>select_words_regex: ^(Jo|Amy|Meg|Beth)\s

input_base_filenames:
  - little_women_part_1
  - little_women_part_2</code></pre>
<p>To get these values from config.yaml into the <strong>all</strong>
rule, I need to introduce the <code>expand</code> function. <a
href="https://snakemake.readthedocs.io/en/stable/snakefiles/rules.html#the-expand-function"
target="_blank">expand</a> is a helper function that comes with
Snakemake. It’s easier to explain if we show a few examples of inputs
and outputs:</p>
<blockquote>
<pre><code>$ python
from snakemake.io import expand
expand(&quot;{dataset}/a.txt&quot;, dataset=[1, 2, 3])
[&#39;1/a.txt&#39;, &#39;2/a.txt&#39;, &#39;3/a.txt&#39;]
.
expand(&quot;{a}.{b}&quot;, a=[&#39;fig1&#39;, &#39;fig2&#39;, &#39;fig3&#39;], b=[&#39;png&#39;, &#39;pdf&#39;])
[&#39;fig1.png&#39;, &#39;fig1.pdf&#39;, &#39;fig2.png&#39;, &#39;fig2.pdf&#39;, &#39;fig3.png&#39;, &#39;fig3.pdf&#39;]
.
expand(&quot;results/{x}.select_words.txt&quot;, \
...         x=[&quot;little_women_part_1&quot;, &quot;little_women_part_2&quot;])
[&#39;results/little_women_part_1.select_words.txt&#39;, &#39;results/little_women_part_2.select_words.txt&#39;]
exit()
$</code></pre>
</blockquote>
<p><code>expand</code> accepts a string pattern with named placeholders
(e.g. {dataset}) and named lists (e.g. dataset=[1, 2]); it builds a list
of strings by inserting the list values into the pattern.</p>
<p>That last example above is exactly what we need in the
<strong>all</strong> rule. The final step is to adapt the example above
to use the config file, like so:</p>
<pre><code>rule all:
    input: expand(&quot;results/{x}.select_words.txt&quot;, \
        x=config[&#39;input_base_filenames&#39;])</code></pre>
<p>The Snakefile is now free of any references to specific inputs. The
workflow can be run on a different set of inputs by using a new
config.yaml.</p>
<pre><code>configfile: &quot;config/config.yaml&quot;

rule all:
    input: expand(&quot;results/{x}.select_words.txt&quot;, \
    x=config[&#39;input_base_filenames&#39;])


rule select_words:
    input: &quot;results/{base_name}.sort_counts.txt&quot;
    output: &quot;results/{base_name}.select_words.txt&quot;
    params: regex=config[&quot;select_words_regex&quot;]
    shell: &quot;egrep &#39;{params.regex}&#39; {input} &gt; {output}&quot;

rule sort_counts:
    input: &quot;results/{base_name}.count_words.txt&quot;
    output: &quot;results/{base_name}.sort_counts.txt&quot;
    shell: &quot;&quot;&quot;
    sort -k1,1nr {input} \
        | awk &#39;BEGIN {{print &quot;word\tcount&quot;}} {{ print $2 &quot;\t&quot; $1}}&#39; \
            &gt; {output}
        &quot;&quot;&quot;
rule count_words:
    input: &quot;results/{base_name}.split_words.txt&quot;
    output: &quot;results/{base_name}.count_words.txt&quot;
    shell: &quot;sort {input} | uniq -c &gt; {output}&quot;

rule split_words:
    input: &quot;inputs/{base_name}.txt&quot;
    output: &quot;results/{base_name}.split_words.txt&quot;
    shell: &quot;cat {input} | tr -cs &#39;[:alpha:]&#39; &#39;\n&#39; &gt; {output}&quot;
</code></pre>
<hr />
</div>
<div id="snakemake-on-great-lakes" class="section level2">
<h2>Snakemake on Great Lakes</h2>
<p>All this time we’ve been executing the workflow on the login-nodes.
It’s generally fine to execute the <code>snakemake</code> process itself
on the login-node, but for a more demanding workflow you should execute
the rules on Great Lakes worker nodes.</p>
<p>It’s straightforward to do this by using a profile. A profile is
essentially just a little config file that tells <code>snakemake</code>
how it should submit rules as jobs. Check out snakemake docs for details
on <a
href="https://snakemake.readthedocs.io/en/stable/executing/cli.html#profiles"
target="_blank">profile</a> options. We’ll start with a simple profile
built for Great Lakes:</p>
<pre class="sh"><code>cp -r /nfs/turbo/umms-bioinf-wkshp/shared-data/profile config/</code></pre>
<div class="wrapper">
<table class="fig">
<tr>
<th class="fig" style="text-align: left">
config/profile/config.yaml
</th>
</tr>
<tr>
<td class="fig">
<pre>
printshellcmds: True
keep-going: True
use-singularity: True
singularity-args: "--cleanenv -B $(pwd)"
cluster: sbatch --job-name=alcott_{rule}_{wildcards} --account=bioinf_wkshp_class --partition=standard --nodes=1 --cpus-per-task={resources.cpus} --mem={resources.mem_mb} --time={resources.time_min} --output=cluster_logs/%x-%j.log --parsable
default-resources: [cpus=1, mem_mb=2000, time_min=120]
rerun-triggers: [mtime]
latency-wait: 300
jobs: 5
</pre>
</td>
</tr>
</table>
</div>
<p><br/></p>
<p>Invoking <code>snakemake</code> with this profile submits each rule a
an sbatch job:</p>
<pre class="sh"><code># clear out existing files so there&#39;s something to do
rm results/*

snakemake --profile config/profile</code></pre>
<p>Note that you can adjust the SLURM job geometry by adding a
<em>resources</em> directive to a rule. These values will be
incorporated into the sbatch request for this rule.</p>
<pre><code>rule select_words:
    input: &quot;results/{base_name}.sort_counts.txt&quot;
    output: &quot;results/{base_name}.select_words.txt&quot;
    resources: cpus=4, mem_mb=4000, time_min=5  
    params: regex=config[&quot;select_words_regex&quot;]
    shell: &quot;egrep &#39;{params.regex}&#39; {input} &gt; {output} #--fake-num-threads={resources.cpus}&quot;</code></pre>
<p><br/></p>
</div>
<div id="using-tmux" class="section level2">
<h2>Using tmux</h2>
<p><a href="https://en.wikipedia.org/wiki/Tmux">tmux</a> is a unix
utility that let’s you create virtual screens inside of a single
session. (It’s like the Unix utility <em>screen</em>, but more powerful
and current.) For our purposes, the killer feature of tmux is captured
in this simple vignette:</p>
<ul>
<li>login to great lakes and start a tmux session on the login-node:
<code>tmux</code></li>
<li>start a long running task (e.g. a snakemake workflow which submits
to the cluster)</li>
<li>detach from that tmux session: hit <code>ctrl-b</code>
<code>d</code></li>
<li>turn off your computer, go home, and login to great lakes from
different computer</li>
<li>reattach to that tmux session: <code>tmux attach</code></li>
<li>you will see your long-running task has been running all along,
completely unaffected by your wanderings</li>
</ul>
<p>And in fact, tmux can actually do much, much more. Here’s a nifty <a
href="https://opensource.com/sites/default/files/gated-content/osdc_cheatsheet-tmux-2021.6.25.pdf"
target="_blank">cheatsheet</a>.</p>
<p>If you start using tmux, keep in mind that when you login into
<code>greatlakes.arc-ts.umich.edu</code> it’s actually forwarding you on
to one of <em>three</em> login nodes: gl-login1, gl-login2, or
gl-login3. (Which one you get is basically random.) This means if you
want to reconnect to a detached tmux session, you have to get to the
right login node.</p>
<p>So, if you start a tmux session, note which login node you are using
using the <code>hostname</code> command:</p>
<blockquote>
<pre><code>$ hostname
gl-login2.arc-ts.umich.edu</code></pre>
</blockquote>
<p>Then when you need to later reconnect to that session, instead of
<code>ssh greatlakes.arc-ts.umich.edu</code> you connect to the specific
hostname you got above:</p>
<blockquote>
<pre><code>ssh gl-login2.arc-ts.umich.edu
# login + Duo
# login banner (blah blah blah)
[cgates@gl-login2 ~]$ tmux ls # list active tmux sessions
0: 1 windows (created Tue Jun 11 20:31:03 2024) [133x25]
[cgates@gl-login2 ~]$ tmux attach # attach to my session</code></pre>
</blockquote>
<p>Voila. You are back in your tmux session, right where you left
it.</p>
<p>Lastly, the scrollbar in the terminal/command window does not work in
tmux. If you need to scroll in tmux, you can hit
<code>ctrl</code>-`<code>b</code> and then <code>[</code> to enter
scroll mode. You can then use cursor or <code>page-up</code>,
<code>page-down</code> to move around the tmux scroll-buffer. Hit
<code>q</code> to quit scoll mode and return to the current prompt.</p>
<hr />
</div>
<div id="pro-tips" class="section level2">
<h2>Pro tips</h2>
<ol style="list-style-type: decimal">
<li>Start learning Snakemake with a simple workflow. Don’t start with a
really complex workflow.</li>
<li>It’s ok to build your Snakefile from the first steps toward the
last, but remember that Snakemake “thinks” from the last output
backward.</li>
<li>Keep your inputs, outputs, and workflow in separate
directories.</li>
<li>Build your Snakefile incrementally.
<ol style="list-style-type: lower-alpha">
<li>Start with a single input.</li>
<li>Add a rule at a time.</li>
<li>Make small changes and use <code>dry-run</code> liberally.</li>
<li>OK to start by hardcoding filenames; put the wildcards in
afterward.</li>
</ol></li>
<li>Keep consistent naming conventions. Including your rule names in
your file names will help you troubleshoot. Some people include numbers
in the file names or place different outputs in different directories.
Choose a pattern that works for you and be consistent.</li>
<li>If you subset your input data so that your steps are fast and light,
you can prototype the Snakefile on the login-nodes. (But be ready to hit
control-C if you wander into a computationally intensive step.)</li>
<li>Use spaces not tabs.</li>
<li>Use config files to avoid hardcoding paths or “magic values” in the
Snakefile.</li>
<li>Checkout Snakemake’s styleguide on <a
href="https://snakemake.readthedocs.io/en/stable/snakefiles/deployment.html#distribution-and-reproducibility">Distribution
and Reproducibility</a>. It has great recommendations on how to organize
files and directories.</li>
</ol>
<hr />
</div>
<div id="key-ideas" class="section level2">
<h2>Key ideas</h2>
<ul>
<li><p>You can visualize the Snakemake DAG or rulegraph using
<code>dot</code>.</p></li>
<li><p>You can make you Snakefile more extensible with:</p>
<ul>
<li>a leading <strong>all</strong> rule to specify the final
outputs</li>
<li><strong>Wildcards</strong> to avoid hardcoded paths and file
names</li>
<li>A <strong>configuration file</strong> to pull buts that change out
of the Snakefile itself</li>
<li><strong>param</strong> directives to fine-tune commands</li>
<li>A <strong>profile</strong> config file and <strong>resource</strong>
directives to right-size resourcing on an HPC</li>
</ul></li>
<li><p><strong>tmux</strong> let’s you run programs that will keep
running if you are disconnected</p></li>
<li><p>Command line flags we reviewed:</p>
<ul>
<li>–dag –rulegraph</li>
<li>–forceall</li>
<li>-p –printshellcmds</li>
<li>–profile</li>
</ul></li>
</ul>
<hr />
</div>
<div id="exercises" class="section level2">
<h2>Exercises</h2>
<div id="exercise-1a" class="section level3">
<h3>Exercise 1A:</h3>
<p>cd to the project publications</p>
<p>Given this DAG can you fill in the missing parts of the Snakefile
below and regenerate a similar DAG figure?</p>
<p><img src="images/advanced_snakemake/exercise1_dag.png" /></p>
<p>FYI:</p>
<ul>
<li>All the rules you need are in place, but most of them have
incomplete input and output directives.</li>
<li>The shell directives in this Snakefile reference “external” scripts
in the workflow scripts directory. You will <em>not</em> need to change
any shell directives for this exercise.</li>
<li>The actual results files for this workflow are merely stubs and not
interesting. Generating the right Snakefile / DAG is the focus.</li>
</ul>
<pre><code>rule all:
    input: &#39;results/knit_paper.html&#39;

rule knit_paper:
    input: &#39;inputs/intro.md&#39;, ??
    output: ??
    shell: &#39;workflow/scripts/knit_paper.sh {input} {output}&#39;

rule make_plot:
    input: ??
    output: ??
    shell: &#39;workflow/scripts/make_plot.sh {input} {output}&#39;

rule make_table:
    input: ??
    output: ??
    shell: &#39;workflow/scripts/make_table.sh {input} {output}&#39;

rule filter:
    input: ??
    output: ??
    params: chromosome=&#39;6&#39;
    shell: &#39;workflow/scripts/filter.sh {input} {output} {params.chromosome}&#39;

rule download_data:
    input: &#39;/nfs/turbo/umms-bioinf-wkshp/workshop/shared-data/agc-endpoint/0042-WS/sample_A.genome.bam&#39;
    output:
    shell: &#39;workflow/scripts/download_data.sh {input} {output}&#39;
</code></pre>
</div>
<div id="exercise-1b" class="section level3">
<h3>Exercise 1B:</h3>
<p>Use params and a config file to extract hardcoded absolute paths and
any other “magic values”.</p>
</div>
<div id="exercise-1c" class="section level3">
<h3>Exercise 1C:</h3>
<p>Extend the workflow to cover more inputs as illustrated in the DAG
below:</p>
<p><img src="images/advanced_snakemake/exercise1C_dag.png" /></p>
</div>
<div id="exercise-1d" class="section level3">
<h3>Exercise 1D:</h3>
<p>Copy the profile/config.yaml from
workflows/project_alcott/alcott_snakemake and force re-run the Snakefile
using worker nodes instead of the login-node.</p>
<hr />
</div>
</div>
<div id="references-and-links" class="section level2">
<h2>References and links</h2>
<ul>
<li><a href="https://snakemake.readthedocs.io/en/stable/index.html"
class="uri">https://snakemake.readthedocs.io/en/stable/index.html</a>
<ul>
<li><a
href="https://snakemake.readthedocs.io/en/stable/executing/cli.html#all-options"
class="uri">https://snakemake.readthedocs.io/en/stable/executing/cli.html#all-options</a></li>
<li><a
href="https://snakemake.readthedocs.io/en/stable/snakefiles/writing_snakefiles.html#grammar"
class="uri">https://snakemake.readthedocs.io/en/stable/snakefiles/writing_snakefiles.html#grammar</a></li>
</ul></li>
</ul>
<hr />
<table style="width:100%;">
<colgroup>
<col width="28%" />
<col width="42%" />
<col width="28%" />
</colgroup>
<thead>
<tr class="header">
<th align="left"><a href="Module_advanced_snakemake.html">Previous
lesson</a></th>
<th align="center"><a href="#top">Top of this lesson</a></th>
<th align="right"><a href="Module_transferring_data_globus.html">Next
lesson</a></th>
</tr>
</thead>
<tbody>
</tbody>
</table>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3,h4",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
